src/main/java/org/commcare/formplayer/session/MenuSession.java

```diff
@@ -258,15 +258,34 @@ public Screen getNextScreen(boolean needsFullEntityScreen, EntityScreenContext e
             queryScreen.init(sessionWrapper);
             return queryScreen;
         } else if (next.equalsIgnoreCase(SessionFrame.STATE_SYNC_REQUEST)) {
-            String username = session.getAsUser() != null ?
-                    StringUtils.getFullUsername(session.getAsUser(), session.getDomain()) : null;
-            FormplayerSyncScreen syncScreen = new FormplayerSyncScreen(username);
-            syncScreen.init(sessionWrapper);
-            return syncScreen;
+            return getSyncScreen();
         }
         throw new RuntimeException("Unexpected Frame Request: " + sessionWrapper.getNeededData());
     }
 
+    /**
+     * Get next screen for the current request, based on the current state of the session,
+     * but only initialize and return the screen if it is of type `FormplayerSyncScreen`.
+     */
+    public FormplayerSyncScreen getNextScreenIfSyncScreen(boolean needsFullEntityScreen, EntityScreenContext entityScreenContext) throws CommCareSessionException {
+        String next = sessionWrapper.getNeededData(sessionWrapper.getEvaluationContext());
+        if (next.equalsIgnoreCase(SessionFrame.STATE_DATUM_COMPUTED)) {
```

Comment:
We need to check for `next == null`

Reply:
I had it at first but i ended up removing it, see this [comment](https://github.com/dimagi/formplayer/pull/1608#discussion_r1705980081). But let me know if you disagree

Reply:
It seems like the tests are failing currently due to `next` being `null`

Reply:
ah yes you're right, thank you [0aa1810](https://github.com/dimagi/formplayer/pull/1608/commits/0aa1810556afd85e075e58f8493d8259d2590939)